// release-please/build/src/strategies/go-librarian.js
"use strict";
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoLibrarian = void 0;
const base_1 = require("./base");
const changelog_1 = require("../updaters/changelog");
const librarian_yaml_1 = require("../updaters/librarian-yaml");
const version_1 = require("../version");
const version_go_1 = require("../updaters/go/version-go");
class GoLibrarian extends base_1.BaseStrategy {
    constructor(options) {
        var _a, _b;
        options.changelogPath = (_a = options.changelogPath) !== null && _a !== void 0 ? _a : 'CHANGES.md';
        super(options);
        this.versionFile = (_b = options.versionFile) !== null && _b !== void 0 ? _b : 'internal/version.go';
    }
    async buildUpdates(options) {
        const updates = [];
        const version = options.newVersion;
        if (!this.skipChangelog) {
            updates.push({
                path: this.addPath(this.changelogPath),
                createIfMissing: true,
                updater: new changelog_1.Changelog({
                    version,
                    changelogEntry: options.changelogEntry,
                }),
            });
        }
        if (this.versionFile) {
            updates.push({
                path: this.addPath(this.versionFile),
                createIfMissing: false,
                updater: new version_go_1.VersionGo({
                    version,
                }),
            });
        }
        updates.push({
            path: 'librarian.yaml',
            createIfMissing: false,
            updater: new librarian_yaml_1.LibrarianYamlUpdater({
                version,
                packagePath: this.path,
                component: this.component,
            }),
        });
        return updates;
    }
    initialReleaseVersion() {
        if (this.initialVersion) {
            return version_1.Version.parse(this.initialVersion);
        }
        return version_1.Version.parse('1.0.0');
    }
}
exports.GoLibrarian = GoLibrarian;
