// release-please/build/src/updaters/librarian-yaml.js
"use strict";
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LibrarianYamlUpdater = void 0;
const default_1 = require("./default");
const yaml = require("yaml");
const logger_1 = require("../util/logger");
/**
 * Updates a librarian.yaml file.
 */
class LibrarianYamlUpdater extends default_1.DefaultUpdater {
    constructor(options) {
        super(options);
        this.specialArtifacts = new Map([
            ['google-cloud-java', 'google-cloud-java'],
        ]);
        this.packagePath = options.packagePath;
        this.component = options.component;
    }
    /**
     * Given initial file contents, return updated contents.
     * @param {string} content The initial content
     * @returns {string} The updated content
     */
    updateContent(content, _logger = logger_1.logger) {
        const doc = yaml.parseDocument(content);
        if (!doc || doc.errors.length > 0) {
            throw new Error(`Invalid yaml, cannot be parsed: ${doc.errors
                .map(e => e.message)
                .join(', ')}`);
        }
        const libraries = doc.get('libraries');
        if (!libraries || !yaml.isSeq(libraries)) {
            return content;
        }
        let modified = false;
        for (const library of libraries.items) {
            if (!yaml.isMap(library))
                continue;
            const libraryJSON = library.toJSON();
            let newVersion = undefined;
            if (this.versionsMap) {
                // Multi-version (Java style)
                const artifactID = this.findArtifactID(libraryJSON);
                if (this.versionsMap.has(artifactID)) {
                    newVersion = this.versionsMap.get(artifactID);
                }
            }
            else {
                // Single version (Go, Python, Node style)
                const isGoMatch = (this.packagePath && libraryJSON.name === this.packagePath) ||
                    (this.component && libraryJSON.name === this.component);
                const isPythonNodeMatch = this.packagePath &&
                    this.deriveOutputDirectory(libraryJSON) === this.packagePath;
                if (isGoMatch || isPythonNodeMatch) {
                    newVersion = this.version;
                }
            }
            if (newVersion) {
                const newVersionStr = newVersion.toString();
                if (library.get('version') !== newVersionStr) {
                    library.set('version', newVersionStr);
                    modified = true;
                }
                if (this.versionsMap) {
                    const isSnapshot = newVersion.preRelease === 'SNAPSHOT';
                    if (!isSnapshot) {
                        let java = library.get('java');
                        if (!yaml.isMap(java)) {
                            const javaNode = doc.createNode({});
                            library.set('java', javaNode);
                            java = javaNode;
                            modified = true;
                        }
                        if (yaml.isMap(java)) {
                            if (java.get('released_version') !== newVersionStr) {
                                java.set('released_version', newVersionStr);
                                modified = true;
                            }
                        }
                    }
                }
            }
        }
        if (modified) {
            return doc.toString({ lineWidth: 0 });
        }
        return content;
    }
    deriveOutputDirectory(library) {
        if (library.output) {
            return library.output;
        }
        return `packages/${library.name}`;
    }
    findArtifactID(library) {
        const artifact = this.specialArtifacts.get(library.name);
        if (artifact) {
            return artifact;
        }
        if (library.java && library.java.artifact_id) {
            return library.java.artifact_id;
        }
        return `google-cloud-${library.name}`;
    }
}
exports.LibrarianYamlUpdater = LibrarianYamlUpdater;
